Brain fixes and changes

phil235 opened the pull request:
- <b>Brains can no longer be blinded, even when inside a MMI.</b> Because they have no eyes. Also because having to handle blindness when moving brain in and out of MMI (making them DEAD or CONSCIOUS) is a pain. Fixes MMI brain becoming forever blind when blinded by a statue.
- Fixes borg made from staff of change not having a brain organ in their MMI. Fixes #15048
- Removes some duplicated procs in mob.dm (probably due to incorrectly fixed merge conflicts)
- You can now attack brains with a weapon even when not inside a MMI (previously only when in MMI).
- Fixes brain vision not being given mech sight when put inside a mech.
- Removing a brain from its MMI place the brain in your hand instead of on the ground.
- Fixes built cyborgs containing two mmis.
- <b>Damaged brain from mob attacks can't be placed inside a cyborg.</b> (Brains can only be damaged by attacks with a weapon).
- Robotized humans (called Robotize()) , such as roundstart cyborgs, now have an MMI with the human's name (instead of having a brain with a borg name, which was inconsistent with built cyborgs).
- Syndiborgs bought by nukeops now have a brain with the same last name as the ops team.
- Fixes brain being killable when inside cyborgs. Fixes borg dropping a dead mmi brain unable to talk. Brains are now immune to environment damage and radiation. Fixes #15312
- Fixes brain being able to suicide multiple times (if put in and out of an MMI)
- Admins can now directly spawn brain mobs. It automatically puts it inside a brain organ. Mostly useful for testing purposes and badminery.
- Fixes defibbed corpse not getting blind overlay despite being in crit.
- Brain mobs that are beaten to death can still be inserted inside a brainless human corpse but defibbing the corpse cannot work (the brain is too damaged). On the other hand cloning that corpse will still work.
- Deffibing a brainless corpse no longer gives "further attempts may be successful".
- Fixes brain mob's container var not being nullified on Destroy().

:cl: phil235
rscadd: Brains can no longer be blinded.
bugfix: Fixes brain being immortal. They are immune to everything but melee weapon attacks. Brains not inside a MMI can now be attacked just like MMIs. Damaged brain can't be put into a robot, they can still be put inside a brainless humanoid corpse and cloned, but deffibing the corpse always fails (can't revive someone with a brain beaten to a pulp).
/:cl:

phil235 committed:
Brains can no longer be blinded, even when inside a MMI.
Fixes borg made from staff of change not having a brain organ in their MMI.
Removes some duplicated procs in mob.dm (probably due to incorrectly fix merge conflicts)
You can now attack brains with a weapon even when not inside a MMI (previously only when in MMI).
Fixes brain vision not being given mech sight when put inside a mech.
Removing a brain from its MMI not place the brain in your hand instead of on the ground.
Fixes built cyborgs containing two mmis.
Damaged brain from attack can't be placed inside a cyborg.
Robotized humans (called Robotize()) , such as roundstart cyborgs, now have an MMI with the human's name (instead of having a brain with a borg name, which was inconsistent with built cyborgs).
Syndiborgs bought by nukeops now have a brain with the same last name as the ops team.
Fixes Borg created with staff of change not containing a brain organ.
Fixes brain being killable when inside cyborgs. Fixes borg dropping a dead mmi brain unable to talk. Brains are now immune to environment damage and radiation.
Fixes brain being able to suicide multiple times (if put in and out of an MMI)

Ultimate-Chimera commented:
:+1:

phil235 commented on code/game/objects/items/robot/robot_parts.dm:
```diff
@@ -210,7 +210,6 @@
 			O.invisibility = 0
 			//Transfer debug settings to new mob
 			O.custom_name = created_name
-			O.updatename("Default")
```
updatename is already called below, and that proc doesn't even use any argument anymore.

phil235 commented on code/modules/mob/transform_procs.dm:
```diff
@@ -320,33 +320,41 @@
+	R.job = "Cyborg"
+	R.notify_ai(1)
+
+	. = R
```
I renamed the robot from O to R, for better code readability.

phil235 committed:
woops, brain still need to die when beaten.

JJRcop commented:
https://www.youtube.com/watch?v=XyFfmGf3b2Y

neersighted commented on code/modules/mob/living/carbon/brain/brain.dm:
```diff
@@ -6,6 +6,15 @@
 	var/timeofhostdeath = 0
 	var/emp_damage = 0//Handles a type of MMI damage
 	has_limbs = 0
+	stat = DEAD //we start dead by default
```
yes thank you this is more sane

neersighted commented on code/modules/mob/living/carbon/brain/life.dm:
```diff
@@ -3,72 +3,20 @@
-		..()
+	return
 
 /mob/living/carbon/brain/handle_environment(datum/gas_mixture/environment)
```
Maybe make brains take damage over time if not chilled? We can give robotics a freezer

phil235 replied:
I prefer to err on the side of safety. So far brain mobs were completely immortal. Let's see how people react to this before adding more possible source of damage.

neersighted commented on code/modules/mob/transform_procs.dm:
```diff
@@ -320,33 +320,41 @@
 		qdel(t)
 
-	var/mob/living/silicon/robot/O = new /mob/living/silicon/robot( loc )
+	var/mob/living/silicon/robot/R = new /mob/living/silicon/robot( loc )
```
Don't space the parens

Dorsisdwarf commented:
>
Robotized humans (called Robotize()) , such as roundstart cyborgs, now have an MMI with the human's name (instead of having a brain with a borg name, which was inconsistent with built cyborgs).

I don't really like this. I liked the idea of factory-made top-of-the-line advanced cyborgs vs Jimkin Anderson's Homebrew Assistant In A Robot

neersighted commented:
But don't you know the **deep lore**? Cyborgs and AI are humans mindwiped and forcibly converted as part of Nanotrasen's testing. This station is their testbed.

RemieRichards commented:
not really mindswapped, the mind stays with the brain, the entire brain is removed! :O

neersighted commented:
Mind**wiped** not **swapped**

phil235 committed:
Fixes defibbed corpse not getting blind overlay despite being in crit.
Brain mobs that are beaten to death can still be inserted inside a brainless human corpse but defibbing the corpse cannot work (the brain is too damaged). On the other hand cloning that corpse will still work.
Deffibing a brainless corpse no longer gives "further attempts may be successful".
Fixes brain mob's container var not being nullified on Destroy().

phil235 commented on code/game/objects/items/robot/robot_parts.dm:
```diff
@@ -187,7 +187,7 @@
 				return
 
-			if(BM.stat == DEAD)
+			if(BM.stat == DEAD || (M.brain && M.brain.damaged_brain))
```
The second part of the check should probably be enough but I'm keeping the brain mob stat == DEAD check just to be safe.

phil235 commented on code/modules/mob/living/carbon/brain/brain.dm:
```diff
@@ -2,16 +2,26 @@
 /mob/living/carbon/brain
 	languages = HUMAN
-	var/obj/item/container = null
+	var/obj/item/device/mmi/container = null
```
container is never something other than a mmi, so might as well make its type a MMI so we don't have to do type checks.

phil235 commented on code/modules/mob/living/carbon/brain/brain_item.dm:
```diff
@@ -11,6 +11,7 @@
 	origin_tech = "biotech=4"
 	attack_verb = list("attacked", "slapped", "whacked")
 	var/mob/living/carbon/brain/brainmob = null
+	var/damaged_brain = 0 //whether the brain organ is damaged.
```
This is to replace the "health < deathlevel" check

phil235 commented on code/modules/mob/living/carbon/brain/brain_item.dm:
```diff
@@ -57,6 +58,10 @@
+/obj/item/organ/internal/brain/attackby(obj/item/O, mob/user, params)
+	user.changeNext_move(CLICK_CD_MELEE)
+	if(brainmob)
+		O.attack(brainmob, user) //Oh noooeeeee
```
So we can beat up the brain whether inside a MMI or not.

phil235 commented on code/modules/mob/living/carbon/brain/life.dm:
```diff
@@ -3,72 +3,24 @@
+		else if(istype(loc, /obj/item/organ/internal/brain))
+			BR = loc
+		if(BR)
+			BR.damaged_brain = 1 //beaten to a pulp
```
not great code but brain mob doesn't have a reference to its brain organ sadly.

phil235 commented on code/modules/mob/living/death.dm:
```diff
@@ -50,7 +49,8 @@
 	weakened = 0
 	sleeping = 0
-	update_sight()
+	blind_eyes(1)
```
For cases where the mob goes from full health to dead directly (like removing its brain). We want them to get the blind overlay when they are deffibed back to alive and unconscious.

phil235 commented on code/modules/projectiles/projectile/magic.dm:
```diff
@@ -164,7 +164,6 @@
 						new_mob.invisibility = 0
 						new_mob.job = "Cyborg"
 						var/mob/living/silicon/robot/Robot = new_mob
-						Robot.mmi = new /obj/item/device/mmi(new_mob)
```
Robot/New() already gives us a mmi, no need to make a new one.

phil235 commented on code/modules/mob/mob.dm:
```diff
@@ -1052,12 +1023,6 @@ var/next_mob_id = 0
-		aiPDA.owner = newname
-		aiPDA.name = newname + " (" + aiPDA.ownjob + ")"
-
-
```
this is from an incorrectly fixed merge conflict I think. procs are duplicates

phil235 commented on code/modules/mob/living/death.dm:
```diff
@@ -50,7 +49,8 @@
 	stunned = 0
 	weakened = 0
 	sleeping = 0
-	update_sight()
```
update_sight() is already called by reset_perspective(null) below

tkdrg commented on code/modules/mob/living/carbon/brain/death.dm:
```diff
@@ -3,7 +3,7 @@
 	stat = DEAD
 
-	if(!gibbed && container && istype(container, /obj/item/device/mmi))//If not gibbed but in a container.
+	if(!gibbed && container)//If not gibbed but in a container.
```
You sure you don't need an `istype(container)` for these checks?

phil235 replied:
see https://github.com/tgstation/-tg-station/pull/15441/files#diff-50f99fd5dc8d192812530b373be2348eR5

tkdrg replied:
Well, that doesn't guarantee it will be an MMI; it just casts it automatically due to *assuming* it is, but it's still dynamically typed. However, we probably want to runtime in that edge case either way, so I guess not checking isn't that big of a deal.
